Documentation/Proposals/NNNN-custom-test-execution-traits.md

```diff
@@ -0,0 +1,376 @@
+/// it runs.
+public protocol CustomTestExecuting: Sendable {
+  /// Execute a function for the specified test and/or test case.
+  ///
+  /// - Parameters:
+  ///   - function: The function to perform. If `test` represents a test suite,
+  ///     this function encapsulates running all the tests in that suite. If
+  ///     `test` represents a test function, this function is the body of that
+  ///     test function (including all cases if it is parameterized.)
+  ///   - test: The test under which `function` is being performed.
+  ///   - testCase: The test case, if any, under which `function` is being
+  ///     performed. This is `nil` when invoked on a suite.
+  ///
+  /// - Throws: Whatever is thrown by `function`, or an error preventing
+  ///   execution from running correctly.
+  ///
+  /// This function is called for each ``Trait`` on a test suite or test
+  /// function which has a non-`nil` value for ``Trait/customTestExecutor-1dwpt``.
+  /// It allows additional work to be performed before or after the test runs.
+  ///
+  /// This function is invoked once for the test its associated trait is applied
+  /// to, and then once for each test case in that test, if applicable. If a
+  /// test is skipped, this function is not invoked for that test or its cases.
+  ///
+  /// Issues recorded by this function are associated with `test`.
+  func execute(_ function: @Sendable () async throws -> Void, for test: Test, testCase: Test.Case?) async throws
+}
+
+public protocol Trait: Sendable {
+  // ...
+
+  /// The type of the custom test executor for this trait.
+  ///
+  /// The default type is `Never`.
+  associatedtype CustomTestExecutor: CustomTestExecuting = Never
+
+  /// The custom test executor for this trait, if any.
+  ///
+  /// If this trait's type conforms to ``CustomTestExecuting``, the default
+  /// value of this property is `self` and this trait will be used to customize
+  /// test execution. This is the most straightforward way to implement a trait
+  /// which customizes the execution of tests.
+  ///
+  /// However, if the value of this property is an instance of another type
+  /// conforming to ``CustomTestExecuting``, that instance will be used to
+  /// perform custom test execution instead.  Otherwise, the default value of
+  /// this property is `nil` (with the default type `Never?`), meaning that
+  /// custom test execution will not be performed for tests this trait is
+  /// applied to.
+  var customTestExecutor: CustomTestExecutor? { get }
+}
+
+extension Trait {
+  // ...
+
+  // The default implementation, which returns `nil`.
+  public var customTestExecutor: CustomTestExecutor? { get }
+}
+
+extension Trait where CustomTestExecutor == Self {
+  // Returns `self`.
+  public var customTestExecutor: CustomTestExecutor? { get }
+}
+
+extension Never: CustomTestExecuting {
+  public func execute(_ function: @Sendable () async throws -> Void, for test: Test, testCase: Test.Case?) async throws
```

Comment:
The extension here can be empty; mentioning that it contains the protocol requirements is redundant.